src/datasets/utils/metadata.py

```diff
@@ -106,6 +117,105 @@ def to_yaml_string(self) -> str:
+                Examples of data_files in YAML:
+
+                   data_files: data.csv
+
+                   data_files: data/*.png
+
+                   data_files:
+                    - part0/*
+                    - part1/*
+
+                   data_files:
+                    - split: train
+                      pattern: train/*
+                    - split: test
+                      pattern: test/*
+                """
+            )
+            if not isinstance(yaml_data_files, (list, str)):
+                raise ValueError(yaml_error_message)
+            if isinstance(yaml_data_files, list):
+                for yaml_data_files_item in yaml_data_files:
+                    if not isinstance(yaml_data_files_item, str) or not (
+                        isinstance(yaml_data_files_item, dict)
+                        and sorted(yaml_data_files_item)
+                        == [
+                            "pattern",
+                            "split",
+                        ]
+                    ):
+                        raise ValueError(yaml_error_message)
+
+    @classmethod
+    def from_metadata(cls, dataset_metadata: DatasetMetadata) -> "MetadataConfigs":
+        if dataset_metadata.get(cls.FIELD_NAME):
+            metadata_configs = dataset_metadata[cls.FIELD_NAME]
+            if isinstance(metadata_configs, dict):  # single configuration
+                if "config_name" not in metadata_configs:
+                    metadata_configs["config_name"] = "default"
+                metadata_configs = [metadata_configs]
+            elif not isinstance(metadata_configs, list):
+                raise ValueError(f"Expected {cls.FIELD_NAME} to be a dict or a list, but got '{metadata_configs}'")
+            for metadata_config in metadata_configs:
+                cls._raise_if_not_valid(metadata_config)
+            return cls(
+                {
+                    config["config_name"]: {param: value for param, value in config.items() if param != "config_name"}
+                    for config in metadata_configs
+                }
+            )
+        return cls()
+
+    def to_metadata(self, dataset_metadata: DatasetMetadata) -> None:
+        if self:
+            for metadata_config in self.values():
+                self._raise_if_not_valid(metadata_config)
+            current_metadata_configs = self.from_metadata(dataset_metadata)
+            total_metadata_configs = dict(sorted({**current_metadata_configs, **self}.items()))
+            if len(total_metadata_configs) > 1:
+                for config_name, config_metadata in total_metadata_configs.items():
+                    config_metadata.pop("config_name", None)
+                dataset_metadata[self.FIELD_NAME] = [
+                    {"config_name": config_name, **config_metadata}
+                    for config_name, config_metadata in total_metadata_configs.items()
+                ]
+            elif len(total_metadata_configs) == 1:
```

Comment:
You can just remove this to always have a list imo :)

Reply:
And this would make the code much readable

Reply:
okay the problem is that `configs` is already used in [`huggingface_hub.DatasetCardData`](https://huggingface.co/docs/huggingface_hub/main/en/package_reference/cards#huggingface_hub.DatasetCardData.configs) and is supposed to be a list of strings. So we have to invent a new name or rollback to `builder_configs`. Or modify the structure on `huggingface_hub`'s side, not sure if it doesn't cause any problems, I don't know how this field is used.

Reply:
and I think we actually need to support a dict in configs yaml to allow for not specifying a config name when there is a single config.

Reply:
then yaml for a single config without a config_name field would look like this:
```
configs:
  - data_dir: v1
    drop_labels: true
```
A bit weird, wdyt?

Reply:
this sounds good to me
```
configs:
  - config_name: default
    data_dir: v1
    drop_labels: true
```

Reply:
so we inforce users to always specify `config_name`?

Reply:
Yes I think we can do that

Reply:
sounds goo, that would make the code more straightforward.
why don't we have the same for `dataset_info`? it allows for both dict and list